test/token/ERC20/extensions/ERC4626.test.js

```diff
@@ -44,6 +45,81 @@ contract('ERC4626', function (accounts) {
     }
   });
 
+  describe('reentrancy', async function () {
+    const reenterType = {
+      Before: 1,
+      After: 2,
+    };
+
+    let amount;
+    let token;
+    let vault;
+
+    beforeEach(async function () {
+      amount = web3.utils.toBN(10).pow(decimals);
+      token = await ERC4626ReentrantAsset.new();
+
+      vault = await ERC4626.new('', '', token.address);
+      await token.$_mint(holder, amount);
+      await token.$_mint(token.address, amount); // Requires balance to reenter
+    });
+
+    // During a `_deposit`, the vault does `transferFrom(depositor, vault, assets)` -> `_mint(receiver, shares)`
+    // such that a reentrancy BEFORE the transfer guarantees the price is kept the same.
+    // If the order of transfer -> mint is changed to mint -> transfer, the reentrancy could be triggered on an
+    // intermediate state in which the ratio of assets/shares has been decreased (more shares than assets).
+    it('reenters before deposit without changing the price', async function () {
+      // Schedules a reentrancy from the token contract
+      const reenterAmount = 1;
+      await token.functionCall(token.address, vault.contract.methods.approve(vault.address, reenterAmount).encodeABI());
+      await token.scheduleReenter(
+        reenterType.Before,
+        vault.address,
+        vault.contract.methods.deposit(reenterAmount, holder).encodeABI(),
+      );
+
+      const sharesBefore = await vault.previewDeposit(amount, { from: holder });
+
+      // A holder approves and deposits as normally, triggering the _beforeTokenTransfer hook
+      await token.approve(vault.address, constants.MAX_UINT256, { from: holder });
+      await vault.deposit(amount, holder, { from: holder });
+
+      // Assert prices is kept
+      const sharesAfter = await vault.previewDeposit(amount, { from: holder });
+      expect(sharesBefore).to.be.bignumber.eq(sharesAfter);
+    });
+
+    // During a `_withdraw`, the vault does `_burn(owner, shares)` -> `transfer(receiver, assets)`
+    // such that a reentrancy AFTER the transfer guarantees the price is kept the same.
+    // If the order of burn -> transfer is changed to transfer -> burn, the reentrancy could be triggered on an
+    // intermediate state in which the ratio of shares/assets has been decreased (more assets than shares).
+    it('reenters after withdraw without changing the price', async function () {
```

Comment:
similar to the other test, I would do it differently

- setup the vault (deposit). Not that you'll either need more user than just holder, of you'll need holder to do a partial withdrawal
- remove the `functionCall` on line 99 and 100
- program the scheduleReenter so that it calls token._burn(vault, some resonable amount) → this is an operation that actually mess up the price.
- do the preview withdraw like you do
- do the withdraw, and check the the event matched the preview
- do another preview withdraw and observe that the price was modified by the reentrancy (but it did not affect the withdraw)